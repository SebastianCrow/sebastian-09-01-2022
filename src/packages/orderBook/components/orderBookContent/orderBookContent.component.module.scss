@import '../../../../styles/layout';

.tablesContainer {
  display: flex;
  height: 100%;
  overflow-y: auto;

  .loadingOverlay {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
  }

  .tablesContent {
    display: flex;
    width: 100%;
    // `align-items: center` is problematic when the content overflows the flex parent (cannot scroll to part of the content).
    // Internal wrapper with `margin: auto` centers the content and makes it always accessible.
    margin: auto;

    @include mobile {
      flex-direction: column-reverse;
    }

    .spreadContainer {
      display: flex;
      align-items: center;
      justify-content: center;
      padding: $spacing;
    }
  }

  @include desktop {
    align-items: center;

    .tableColumn {
      flex: 1;
    }
  }
}
