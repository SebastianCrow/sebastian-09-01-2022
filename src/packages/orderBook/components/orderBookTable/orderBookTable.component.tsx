import React, { FunctionComponent } from 'react';
import { Table } from '../../../ui/table/table.component';
import { useComputeOrderBookTableData } from '../../hooks/useComputeOrderBookTableData.hook';
import { ComputedPriceInfo, PriceDataType } from '../../state/orderBook.types';

interface OrderBookTableProps {
  priceInfoList: ComputedPriceInfo[];
  priceDataType: PriceDataType;
}

export const OrderBookTable: FunctionComponent<OrderBookTableProps> = ({
  priceInfoList,
  priceDataType,
}) => {
  const { columns, data, options } = useComputeOrderBookTableData({
    priceInfoList,
    priceDataType,
  });
  return <Table columns={columns} data={data} options={options} />;
};
