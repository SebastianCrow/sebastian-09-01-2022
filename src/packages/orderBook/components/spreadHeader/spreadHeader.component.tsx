import React, { FunctionComponent } from 'react';
import { Text } from '../../../ui/text/text.component';
import { useSelectSpreadInfo } from '../../hooks/useSelectOrderBookState.hook';

export const SpreadHeader: FunctionComponent = () => {
  const spread = useSelectSpreadInfo();
  // TODO: Placeholder for undefined spread
  if (!spread) {
    return null;
  }
  return (
    <Text color="secondary">{`Spread: ${spread.value} (${spread.percent}%)`}</Text>
  );
};
