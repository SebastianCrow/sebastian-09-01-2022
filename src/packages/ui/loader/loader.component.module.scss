@import '../../../styles/variables';
@import '../../../styles/themes/dark';

.container {
  padding: 4 * $spacing;

  .loader {
    color: $color-action !important;
  }
}
