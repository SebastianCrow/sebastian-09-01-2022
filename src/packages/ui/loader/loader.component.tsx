import React, { FunctionComponent } from 'react';
import { CircularProgress } from '@mui/material';
import styles from './loader.component.module.scss';

export const Loader: FunctionComponent = () => {
  return (
    <div className={styles.container}>
      <CircularProgress className={styles.loader} />
    </div>
  );
};
