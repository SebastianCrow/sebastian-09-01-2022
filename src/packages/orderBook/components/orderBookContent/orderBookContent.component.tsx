import React, { FunctionComponent } from 'react';
import cns from 'classnames';
import { SpreadHeader } from '../spreadHeader/spreadHeader.component';
import styles from './orderBookContent.component.module.scss';
import { Loader } from '../../../ui/loader/loader.component';
import { OrderBookTable } from '../orderBookTable/orderBookTable.component';
import { useSelectOrderBookPrices } from '../../hooks/useSelectOrderBookState.hook';

export const OrderBookContent: FunctionComponent = () => {
  const { bids, asks } = useSelectOrderBookPrices();

  const hasData = Boolean(bids && asks);

  return (
    <div className={styles.tablesContainer}>
      {!hasData && (
        <div className={styles.loadingOverlay}>
          <Loader />
        </div>
      )}
      {hasData && bids && asks && (
        <div className={styles.tablesContent}>
          <div className={styles.tableColumn}>
            <OrderBookTable priceInfoList={bids} priceDataType="bids" />
          </div>
          <div className={cns(styles.spreadContainer, 'mobile')}>
            <SpreadHeader />
          </div>
          <div className={styles.tableColumn}>
            <OrderBookTable priceInfoList={asks} priceDataType="asks" />
          </div>
        </div>
      )}
    </div>
  );
};
