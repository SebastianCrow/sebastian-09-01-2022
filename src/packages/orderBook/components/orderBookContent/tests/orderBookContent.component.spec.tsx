import React from 'react';
import { createRendererWithOrderBookStore } from '../../../tests/utils/createRendererWithOrderBookStore.util';
import { OrderBookContent } from '../orderBookContent.component';
import { OrderBookTable } from '../../orderBookTable/orderBookTable.component';
import { SpreadHeader } from '../../spreadHeader/spreadHeader.component';
import { createRendererWithStore } from '../../../../../tests/utils/createRendererWithStore.util';
import { getOrderBookMockStore } from '../../../../../tests/mocks/getOrderBookMockStore';
import { Loader } from '../../../../ui/loader/loader.component';

describe('orderBookContent.component', () => {
  test('renders', () => {
    const renderer = createRendererWithOrderBookStore(<OrderBookContent />);

    expect(renderer.toJSON()).toMatchSnapshot();
    expect(renderer.root.findAllByType(OrderBookTable)).toHaveLength(2);
    // TODO: Visual testing checking media queries (`mobile` and `desktop` classes) to check <SpreadHeader /> display
    expect(renderer.root.findAllByType(SpreadHeader)).toHaveLength(1);
  });

  test('renders loading overlay for `subscribing` state', () => {
    const renderer = createRendererWithStore(
      <OrderBookContent />,
      getOrderBookMockStore({
        connectionStatus: 'subscribing',
      })
    );

    expect(renderer.toJSON()).toMatchSnapshot();
    expect(renderer.root.findAllByType(Loader)).toHaveLength(1);
    expect(renderer.root.findAllByType(OrderBookTable)).toHaveLength(0);
    expect(renderer.root.findAllByType(SpreadHeader)).toHaveLength(0);
  });

  test('does not render content for missing data', () => {
    const renderer = createRendererWithStore(
      <OrderBookContent />,
      getOrderBookMockStore({
        connectionStatus: 'unsubscribed',
        prices: undefined,
      })
    );

    expect(renderer.toJSON()).toMatchSnapshot();
    expect(renderer.root.findAllByType(Loader)).toHaveLength(0);
    expect(renderer.root.findAllByType(OrderBookTable)).toHaveLength(0);
    expect(renderer.root.findAllByType(SpreadHeader)).toHaveLength(0);
  });
});
