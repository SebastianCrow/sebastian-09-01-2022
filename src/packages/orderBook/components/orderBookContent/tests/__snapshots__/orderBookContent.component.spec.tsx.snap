// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderBookContent.component renders 1`] = `
<div
  className="tablesContainer"
>
  <div
    className="tablesContent"
  >
    <div
      className="tableColumn"
    >
      <div
        className="loadingOverlay"
      >
        <span
          className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary loader css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style={
            Object {
              "height": 40,
              "width": 40,
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={Object {}}
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="spreadContainer mobile"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 secondary css-ahj2mt-MuiTypography-root"
      >
        Spread: 
        17.00
         (
        0.05
        %)
      </p>
    </div>
    <div
      className="tableColumn"
    >
      <div
        className="loadingOverlay"
      >
        <span
          className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary loader css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style={
            Object {
              "height": 40,
              "width": 40,
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={Object {}}
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`orderBookContent.component renders loading overlay 1`] = `
<div
  className="tablesContainer"
>
  <div
    className="loadingOverlay"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary loader css-18lrjg1-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
  </div>
</div>
`;
